### DISCUSSION TITLE
How do I give a prerender function an "initial output"? (Build Output API)

### DISCUSSION THREAD

**Discussion by AaronBeaudoin** [vercel/vercel#8289] [Help] [answered]
**Update**: _For a better explanation of what I'm trying to achieve, see [my comment](https://github.com/vercel/vercel/discussions/8289#discussioncomment-3316918) below._

**Update 2**: _This is **possible!!** See my [attempt](https://github.com/vercel/vercel/discussions/8289#discussioncomment-3379166) at a detailed explanation below. Thanks to @TooTallNate for updating the Build Output API ISR [example](https://github.com/vercel/examples/tree/main/build-output-api/on-demand-isr) to help demonstrate all this in action._

### Original Text

I would like to serve my static files as the initial output of my prerender function, and then use the prerender function only to update the pages later after my deploy.

<hr>

Suppose I have a prerender function simply at `/`.

Right now, after a deployment, it's like this:
1. Go to `/`. The prerender function runs. I either have to wait, or I see the fallback page I specified.

What I want is this after a new deployment:
1. Go to `/`. I see a `index.html` I deployed as a static file. It is treated essentially as the "initial output" for my prerender function.
2. If I need to update my `/` page, I use on-demand ISR via `bypassToken` and a separate function to revalidate the page.
3. Now I go to `/` again. I see the new output from my prerender function.

Basically, I want to give my prerender function an initial output so there is no wait time on the first request.

<hr>

_This is different from using a `.vercel/output/functions/<name>.prerender-fallback.<ext>` file because it would treated as an initial **cached** version of the function output (meaning the function wouldn't need to run), and it would depend on the path the visitor is accessing to choose what file to give them. When using a `prerender-fallback` file, the visitor always just gets that one same file regardless of what path they used to get to the prerender function, and the function always runs because the fallback is not considered an initial "output" of the function. Correct me if I'm misunderstanding that._

**Comment by amyegan** [comment]
Thanks for reaching out @AaronBeaudoin. Can you share some more details about your project (or the repo if it's public)? I think the right answer here is going to depend heavily on what frameworks or tools you're already using.

**Reply by AaronBeaudoin** [reply]
**TLDR**: _I need to be able to give a [prerender function](https://vercel.com/docs/build-output-api/v3#vercel-primitives/prerender-functions) with an [`expiration`](https://vercel.com/docs/build-output-api/v3#vercel-primitives/prerender-functions:~:text=Description-,expiration,-Integer%20%7C%20false) of `false` an "initial output" at build time so it doesn't need to run on the first page visit from a user. In this way, it becomes functionally equivalent to a static file (from a visitor performance perspective), except that by using my [`bypassToken`](https://vercel.com/docs/build-output-api/v3#vercel-primitives/prerender-functions:~:text=the%20same%20time.-,bypassToken,-String) I can [revalidate it](https://vercel.com/docs/build-output-api/v3#features/on-demand-incremental-static-regeneration-isr) to "update the static file" (so to speak) without redeploying my entire site. See detailed explanation below._

<hr>

Hi @amyegan, thanks for reaching out!

I don't have a repo to share at the moment. Problem is, I'm trying to describe a feature that as far as I can tell doesn't exist. I think the better approach here would be to try and describe an example use case.

I'm trying to do this in a really framework-agnostic way. I don't use Next.js, and I'm moving away from Nuxt.js in favor of [`vite-plugin-ssr`](https://vite-plugin-ssr.com) (and later even maybe my own low-level Vite SSR custom implementation). The Build Output API is perfect for this because it seems very framework-agnostic to me. **Great work on it so far!** 👍

## Example Use Case

For me, the appeal of using Build Output API [prerender functions](https://vercel.com/docs/build-output-api/v3#vercel-primitives/prerender-functions) (AKA ISR) is the use case where I set the `expiration` parameter to `false` and only ever invalidate the cache on-demand using the `bypassToken`, as described [here](https://vercel.com/docs/build-output-api/v3#features/on-demand-incremental-static-regeneration-isr). I can therefore have pretty much a _static site_ (from a performance perspective). Rather than redeploying my entire site when something changes in my CMS, I can write my own custom triggers in my CMS to call a serverless function on Vercel and only update the page(s) that need to change via my prerender function with on-demand ISR.

<hr>

Take for example, a simple use case such as a _blog_. In fact, you guys have an example of that [here](https://github.com/vercel/examples/tree/main/build-output-api/prerender-functions). In your example:
- There are two blog posts that are prerendered during the build process and deployed as static files.
- The routing config routes to the prerender function after first [prioritizing the filesystem](https://github.com/vercel/examples/blob/main/build-output-api/prerender-functions/.vercel/output/config.json#L4) (static files).
- The routing config causes [multiple paths](https://github.com/vercel/examples/blob/main/build-output-api/prerender-functions/.vercel/output/config.json#L5) (any matching `/blog/<slug>`) to be sent to the prerender function.
- The prerender function has an [`expiration`](https://github.com/vercel/examples/blob/main/build-output-api/prerender-functions/.vercel/output/functions/blog/post.prerender-config.json#L2) of `5` seconds, useful for the example.
- The prerender function has a [`fallback`](https://github.com/vercel/examples/blob/main/build-output-api/prerender-functions/.vercel/output/functions/blog/post.prerender-config.json#L6) [file](https://github.com/vercel/examples/blob/main/build-output-api/prerender-functions/.vercel/output/functions/blog/post.prerender-fallback.html) specified, to be served for the initial prerender.

In my "real-world" case, I would start by modifying the example as follows:
- I would set the prerender function to instead have an `expiration` of `false`.
- I would set the prerender function to have no `fallback` file.

<hr>

Ok, with that "setup" out of the way, here is the feature I'm missing.

I want to prerender some of my blog posts at build time and deploy them as static files, just like the example, but rather than serving them _instead of_ my prerender function, I just want to treat them as the _initial output_ of my prerender function.

In my case, I'm going to revalidate the prerender function manually when data changes in my CMS, so why would I prerender some blogs posts (the static files) at build time, that I can never update without another redeploy?

The `fallback` parameter also isn't helpful here because the prerender function _still runs_ when showing the fallback. It is not mean't to be treated as an _initial output_ for the prerender function. In my case, I would want to provide a blog post file that would be served instead of my prerender function and cause a "cache hit" (in that way it is treated as an "output" of my function); the prerender function would not even run until I later manually trigger it by revalidating it via on-demand ISR, at which point that new output would replace my "initial output". For that reason, my blog post file would also need to unique to the URL that the prerender function is being called from, since my prerender function can render different blog posts depending on what query parameters it is being called with.

Basically, you can summarize what I'm needing as the ability to set an _initial output_ for the prerender function at build time. Why would I (or anyone) want to do this? Well, in the example above, we have some "popular" blog posts that we want to maintain the highest performance for. What if I make an update to one of my "popular" blog posts in my CMS? Well, since they're deployed as static files, and the routing rules prioritize them over the prerender function, it is impossible to change them without a full redeploy. That kind of defeats the purpose of a prerender function (ISR).

## The Core Goal

So, imagine if you will, the following scenario (what I'd like to achieve):
- The routing rules prioritize the prerender function instead of the static files.
- If a static file exists for a given path (AKA I prerendered it myself at build time), it is cached as the function's "output" for the current path when the function is called instead of running the function.
- When I revalidate the prerender function at that path, it is like "replacing" the static file.

Or, alternatively, you can think of it like this:
- The routing rules are the same as they are now.
- No static blog post files at all are deployed from my build step.
- Via some Build Output API config parameter that currently doesn't exist, I give the prerender function a list of query parameter combinations (paths) to prerender at **_build time_**, without needing a user to visit the page.

## The Implications

Think about this for a second. It means that I could deploy a site, and all user visits to my critical pages (including the first users to visit those pages) could have the same performance as a _fully static_ site, but rather than needing a full redeploy to update when a change happens in my CMS, I would be able to update only one or a few pages as needed by revalidating them using on-demand ISR. That's a **powerful** setup. 💪

**Please do not hesitate** to reach out for more clarification if it still isn't clear what I'm asking for. I think this feature would really cement Vercel's ISR functionality through the Build Output API as the _ultimate_ site deployment architecture.

**Reply by AaronBeaudoin** [reply]
## Example Config

An `example.prerender-config.json` for a function at `prerender.func` might look like this (same as the [example](https://vercel.com/docs/build-output-api/v3#vercel-primitives/prerender-functions/configuration/config-example) in the docs, but with a proposed `prerenderBeforeDeploy` parameter at the end):

```jsonc
{
  "expiration": 60,
  "group": 1,
  "bypassToken": "03326da8bea31b919fa3a31c85747ddc",
  "fallback": "example.prerender-fallback.html",
  "allowQuery": ["id"],

  // This is the config parameter I wish existed.
  "prerenderBeforeDeploy": [
    { "id": 1 },
    { "id": 2 },
    { "id": 3 }
  ]
}
```

With this configuration, the prerender function "output" a user would get by going to `/prerender?id=1` would have already been prerendered ahead of time, and they would get a cache hit even on the first visit. However, unlike a static file, I could still update the output without a full redeploy by using on-demand ISR to invalidate the function (by calling it with a `x-prerender-revalidate: <bypassToken>` header). That would be **sweet**.

_Also, the parameter obviously wouldn't have to be called `prerenderBeforeDeploy`. That's just a quick and dirty name I picked for the sake of example. This feature, in combination with fixing the descrepancy mentioned in [#8329](https://github.com/vercel/vercel/discussions/8329#discussioncomment-3337046), would make prerender functions crazy powerful and super intuitive (in my humble opinion)._

**Reply by amyegan** [reply]
Sorry for my delayed response. Wanted to let you know the Build engineers are taking a look. Quick assessment from them is that this kind of thing should already be possible with the Build Output API, but we don't have an example of it available yet. Your [tl;dr](https://github.com/vercel/vercel/discussions/8289#discussioncomment-3316918) seems to be good outline of what you'll need to do to make it work if you want to take a stab at it before we can get an example built.

Thank you for bringing up such a detailed explanation of what you would like to be able to accomplish!

**Reply by AaronBeaudoin** [reply]
Thanks @amyegan, but I have to be honest I'm not sure what you're suggesting. I took a very long "stab at it" before I ever started this discussion, and wasn't able to make it work, otherwise I would have never written anything.

Looking forward to seeing the example. Our team is trying to migrate our site to Vercel and this scenario is the one big roadblock that's keeping us from moving forward.

**Reply by TooTallNate** [reply]
Hi @AaronBeaudoin, we have a pull request open on our examples repo that I believe demonstrates what you are trying to accomplish. Please take a look and let us know if that satisfies your needs: https://github.com/vercel/examples/pull/374

Cheers!

**Reply by AaronBeaudoin** [reply]
I just deployed that example and played around with it, and it appears to doing what I need! **Thanks so much!!!**

If you don't mind, I do have 4 important questions/points:
1. The principle here is something I would have _never_ been able to just guess by playing around with the Build Output API—using a symlink to act as a "function" for each build-time pre-rendered route, and giving each symlink it's own config file and `fallback`. This definitely deserves an explanation in the docs.
2. The docs say that _when the fallback file is served, the Serverless Function will also be invoked "out-of-band" to re-generate a new version of the asset that will be cached and served for future HTTP requests._ Based off of this example, it appears that this is actually _only_ true when `expiration` is _not_ `false`. Am I understanding that correctly? If so, that part of the docs could really use an extra sentence to clarify how that works.
3. Naming the "base" function `[slug].func` had me hung up for a bit because I though `[slug]` was some sort of filesystem routing magic like what Next/Nuxt does. But it looks like in this case that's just the name of the "file". A bit confusing. Maybe rename it to `default.func`, or `base.func`, or even just `index.func` for the example?
4. When I deploy the full example, you can see that Vercel tells me that 6 serverless functions were created. But wait a sec... didn't I actually have 4 functions, with 2 pre-rendered routes linking back to one of them? Did that `[post].func` actually get duplicated 3 times due to my symlinks, or is it still really just one function and this build log output is just misleading? If I wanted to pre-render 100 routes at build time this way, would I be creating 100 extra functions? Does that affect costs?

![image](https://user-images.githubusercontent.com/26592486/184225012-427c9148-9b61-46f8-b405-daefc7687166.png)

**Comment by AaronBeaudoin** [comment, marked as answer]
In case anyone else comes here with the same question, I'd like to do my best to try and answer it clearly here.

@TooTallNate has been kind enough to update the Build Output API ISR [example](https://github.com/vercel/examples/tree/main/build-output-api/on-demand-isr) to demonstrate how this can be accomplished. Although the example is great, the whole design is rather unintuitive, so here's an explanation of how it works.

<hr>

First, we create a normal pre-render function as described [here](https://vercel.com/docs/build-output-api/v3#vercel-primitives/prerender-functions) in the docs, with an associated config in which `expiration` is `false`. We can _optionally_ omit the `fallback` option, in which case every unique route recognized by the pre-render function will be generated on the fly when it is first called. This is pretty well documented. When you _do_ provide a `fallback` file, however, the docs fail to describe how the purpose of the file is different depending on how you're set up.

- If your `expiration` is set to `false` like in this example, the `fallback` file is **always** served until you manually revalidate the route. This is contrary to what the docs describe [here](https://vercel.com/docs/build-output-api/v3#vercel-primitives/prerender-functions/configuration/fallback-static-file:~:text=When%20the%20fallback%20file%20is%20served%2C%20the%20Serverless%20Function%20will%20also%20be%20invoked%20%22out%2Dof%2Dband%22%20to%20re%2Dgenerate%20a%20new%20version%20of%20the%20asset%20that%20will%20be%20cached%20and%20served%20for%20future%20HTTP%20requests.), which only applies...
- ...when your `expiration` is set to a number, in which case the `fallback` file is only served as a "stale" initial response while the pre-render function renders the route in the background.

But what if you want to pre-render some specific routes at build-time for a function that's supposed to handle multiple routes? This is where the provided example is critical for demonstration, because there is no information whatsoever in the docs to help with this scenario.

To do this, create a **symlink** file that points back to your pre-render function, whose path can be parsed by the function in the case that it ever need to be revalidated. It's a bit confusing, because normally your function is either grabbing data from query parameters or from the `x-now-route-matches` in the case that you rewrote some other URL to it, but in this case you'll be manually parsing data out from the URL. This is because the symlink is acting as a "pseudo function" which is the same as the original pre-render function but has a **different URL** when it is called.

Basically, your function is going to need to be flexible with regards to how it gets its data. So, for some example function `example.func` at the root of the filesystem:

- If the data is in query parameters, the function was called directly like `/example?some=data`.
- If the data is in `x-now-route-matches`, the function was rewritten to from some other URL.
- If the data is in the path, the function was called via a symlink.

In the case that you're using a symlink, you'll also need to create a `.prerender-config.json` for your symlink with a `fallback` option specified. In this case, `fallback` has a new **3rd purpose** different from the two described above. In this case, it is acting semantically as the **build-time pre-rendered output** of the pre-render function for that particular route.

<hr>

So in conclusion, your file structure would be something like this:

```sh
.vercel/
  config.json
  output/
    functions/
      prerender/

        # Our "base" pre-render function.
        index.func/

        # Our "base" config with `expiration: false` and `fallback: "index.prerender-fallback.html"`.
        index.prerender-config.json

        # Our "base" fallback which acts as "404" page.
        index.prerender-fallback.html

        # An example of route we pre-render at build time.
        route-1.func@ -> index.func/

        # The config for our pre-rendered route.
        # Has the same config as above except `fallback: "route-1.prerender-fallback.html"`.
        route-1.prerender-config.json

        # Our build time pre-rendered file to respond with.
        route-1.prerender-fallback.html
```

And our `config.json` looks like this:

```json
{
  "version": 3,
  "routes": [
    { "handle": "filesystem" },
    {
      "src": "/prerender/(?<route>[^/]+)(?:/)?",
      "dest": "/prerender?route=$route"
    }
  ]
}
```

This works because `{ "handle": "filesystem" }` causes our `route-1.func` symlink to be matched by `/prerender/route-1` before the rule. So in this case, immediately after a deploy:

1. When going to `/prerender/route-1`, we get the `route-1.prerender-fallback.html` file.  
   The `x-vercel-cache` header has a value of `PRERENDER`.  
   The pre-render function did not run in the background.
2. If we refresh, we get the `route-1.prerender-fallback.html` file again.  
   The `x-vercel-cache` header has a value of `HIT`.  
   The pre-render function still did not run the background.
3. When going to `/prerender/route-2`, we get the `index.prerender-fallback.html` file.  
   The `x-vercel-cache` header has a value of `PRERENDER`.  
   The pre-render function did not run in the background.
4. If we refresh, we get the `index.prerender-fallback.html` file again.  
   The `x-vercel-cache` header has a value of `HIT`.  
   The pre-render function still did not run the background.

So whether you want to **revalidate** `/prerender/route-1` or **generate** `/prerender/route-2`, you'll need to use `bypassToken` to trigger a revalidation via a separate custom handler function.

Alternatively, you can omit `index.prerender-fallback.html` and the corresponding `fallback` option in `index.prerender-config.json`, in which case (1) and (2) will still be true, but:

- When going to `/prerender/route-2`, the pre-render function runs.
  The `x-vercel-cache` header has a value of `MISS`.
- If we refresh, we get the pre-rendered output again.
  The `x-vercel-cache` header has a value of `HIT`.

Time to go into developer hibernation and recover from this.
